[DPE-5847] [Docs] Tutorial rewrite (draft) (#98)
canonical/spark-k8s-bundle: releases/3.4/yaml/docs/tutorial/t-data-processing.md

# 3. Distributed data processing

In this section, you will learn how to use PySpark and Spark Submit to run your Spark jobs. Make sure to finish setting up the environment from the [Environment setup](t-setup.md) page.

## PySpark shell

The `spark-client` snap comes with a built-in Python shell where we can execute commands interactively against an Apache Spark cluster using the Python programming language.

To proceed, run the PySpark interactive CLI:

```bash
spark-client.pyspark --username spark --namespace spark
```

Once the shell is open and ready, you should see a welcome screen similar to the following:

```
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /__ / .__/\_,_/_/ /_/\_\   version 3.4.2
      /_/

Using Python version 3.10.12 (main, Jan 17 2025 14:35:34)
Spark context Web UI available at http://10.181.60.89:4040
Spark context available as 'sc' (master = k8s://https://10.181.60.89:16443, app id = spark-79ab7a21242c4a81bc88c1f71348c102).
SparkSession available as 'spark'.
>>>
```

When you open the PySpark shell, Charmed Apache Spark spawns a couple of executor K8s pods in the background to process commands. 

> You can see them by fetching the list of pods in the `spark` namespace in a separate shell:

Reviewer: ```suggestion
You can see them by fetching the list of pods in the `spark` namespace in a separate host shell:
```

```python
from operator import add
from pyspark.sql import SparkSession

def contains_ubuntu(text):
    return 1 if isinstance(text, str) and "ubuntu" in text.lower() else 0

# Create a Spark session 
spark = SparkSession\
        .builder\
        .appName("CountUbuntuTweets")\
        .getOrCreate()

# Read the CSV file into a DataFrame and convert to RDD
rdd = spark.read.csv("s3a://spark-tutorial/twitter.csv", header=True).rdd

# Count how many rows contain the word "ubuntu" in the "text" column
count = (
    rdd.map(lambda row: row["text"])  # Extract the "text" column
        .map(contains_ubuntu)  # Apply the contains_ubuntu function
        .reduce(add)  # Sum all the 1 and 0 to get the total number of matches
)

# Print the result
print(f"Number of tweets containing Ubuntu: {count}")

spark.stop()
```

We’ve added a few more lines to what we’ve executed so far. The Apache Spark session, which would be available by default in a PySpark shell, needs to be explicitly created. Also, we’ve added `spark.stop()` at the end of the file to stop the Apache Spark session after completion of the job.

> Let’s save the aforementioned script in a file named `count-ubuntu.py` and proveed further to run it.

Reviewer: I think there's a typo here in "and proveed further to run it" but I'm not sure what the intended meaning was.

### Run

When submitting a Spark job, the driver won’t be running in the local machine but on a K8s pod, hence the script needs to be downloaded and then executed remotely in a dedicated pod.
For that reason, we'll copy the file to the S3 storage to be easily accessible from K8s pods.

Upload the file to the Multipass VM:

```bash
multipass transfer count-ubuntu.py spark-tutorial:count-ubuntu.py
```

where `spark-tutorial` is the name of the VM.

Copy the file to the S3-compatible storage:

```bash
aws s3 cp count-ubuntu.py s3://spark-tutorial/count-ubuntu.py
```

Now run the script by issuing the following command:

```bash
spark-client.spark-submit \
    --username spark --namespace spark \
    --deploy-mode cluster \
    s3a://spark-tutorial/count-ubuntu.py
```

When you run the command, you’ll see log output in the console with information about the state of the pods executing the task. 

> While `spark-submit` command spins up K8s pods and runs the script, you can check the K8s pods statuses by running the following command in a different shell on the VM:

Reviewer: ```suggestion
While the `spark-submit` command spins up K8s pods and runs the script, you can check the K8s pods statuses by running the following command in a different shell on the VM:
```

# 2. Distributed data processing

In this section, you will learn how to use PySpark and Spark Submit to run your Spark jobs. Make sure to finish setting up the environment from the [Environment setup](t-setup.md) page.

## PySpark shell

The `spark-client` snap comes with a built-in Python shell where we can execute commands interactively against an Apache Spark cluster using the Python programming language.

> To proceed, run the PySpark interactive CLI:

Author: ```suggestion
To proceed, run the PySpark interactive CLI, specifying the service account to be used for retrieving configurations and running the executor pods (see more information below)
```

pysparkshell-xxxxxxxxxxxxxxxx-exec-2              1/1     Running            0          5s
```

As you can see, PySpark spawned two executor pods within the `spark` namespace. This is the namespace that we provided as a value to the `--namespace` argument when launching `spark-client.pyspark`. It's in these executor pods that data is cached and the computation will be executed, therefore creating a computational architecture that can horizontally scale to large datasets ("big data"). 

On the other hand, the PySpark shell started by the `spark-client` snap will act as a `driver`, controlling and orchestrating the operations of the executors. More information about the Apache Spark architecture can be found in the [Apache Spark documentation](https://spark.apache.org/docs/latest/cluster-overview.html).

The PySpark shell is just like a regular Python shell with Apache Spark Context that can be easily accessed with variables `sc` and `spark` respectively. You can even see this printed in its welcome screen.

### Hello world

Let's start by printing a simple 'hello, world!', just like you'd do in a regular Python shell:

```python
print('hello, world!')
```

> You should see the phrase printed immediately.

Author: Deleted in https://github.com/canonical/spark-k8s-bundle/pull/98/commits/cbb31531bc3e76658ee898a827c1cffbe5d637e5

```

Now the dataset is stored in our MinIO instance: in the `spark-tutorial` bucket with the filename `twitter.csv`.
You can check this by listing all files in the bucket:

```bash
aws s3 ls spark-tutorial
```

#### Distributed dataset processing

To demonstrate how simple it is to use distributed data processing in Charmed Apache Spark, let's count the number of tweets that mention Ubuntu.
Start the PySpark:

```bash
spark-client.pyspark --username spark --namespace spark
```

For distributed and parallel data processing Apache Spark actively uses the concept of a [resilient distributed dataset (RDD)](https://spark.apache.org/docs/latest/rdd-programming-guide.html#resilient-distributed-datasets-rdds), which is a fault-tolerant collection of elements that can be operated on in parallel across the nodes of the cluster.

Read CSV from S3 and create an RDD from our sample dataset:

```python
rdd = spark.read.csv("s3a://spark-tutorial/twitter.csv", header=True).rdd
```

Now that RDD can be used for parallel processing by multiple Apache Spark executors.

Count the number of tweets (lines in CSV) with "text" field containing "Ubuntu" in a case insensitive way:

```python

> count = rdd.map(lambda row: row['text']).map(lambda cell: 1 if isinstance(cell, str) and "ubuntu" in cell.lower() else 0).reduce(add)

Author: Good catch. It looks faster and simpler.
But what I tried to do here is to show an explicit MapReduce with some kind of processing (even if it's just a simple summation). Like a simplified model of a real Big data application.
The `count` does the "reduce" phase implicitly, which is less illustrative as a model.
What would you prefer here? A simpler `count` or explicit `reduce(add)` with 1s and 0s?

spark.stop()
```

We’ve added a few more lines to what we’ve executed so far. The Apache Spark session, which would be available by default in a PySpark shell, needs to be explicitly created. Also, we’ve added `spark.stop()` at the end of the file to stop the Apache Spark session after completion of the job.

Let’s save the aforementioned script in a file named `count-ubuntu.py` and proceed further to run it.

### Run

When submitting a Spark job, the driver won’t be running in the local machine but on a K8s pod, hence the script needs to be downloaded and then executed remotely in a dedicated pod.
For that reason, we'll copy the file to the S3 storage to be easily accessible from K8s pods.

Upload the file to the Multipass VM:

```bash
multipass transfer count-ubuntu.py spark-tutorial:count-ubuntu.py
```

where `spark-tutorial` is the name of the VM.

Copy the file to the S3-compatible storage:

```bash
aws s3 cp count-ubuntu.py s3://spark-tutorial/count-ubuntu.py
```

Now run the script by issuing the following command:

```bash
spark-client.spark-submit \
    --username spark --namespace spark \
    --deploy-mode cluster \
    s3a://spark-tutorial/count-ubuntu.py
```

> When you run the command, you’ll see log output in the console with information about the state of the pods executing the task.

Author: ```suggestion
The --deploy-mode cluster option tells Spark Submit to run the driver on a K8s pod.

When you execute the command, the console will display log output showing the state of the pods running the task.
```
